### PR TITLE
Reduce docker image size

### PR DESCRIPTION
**Pull request by wagoodman**
**Thanks for contributing to Prefect!**

Please describe your work and make sure your PR:

- [x] adds new tests (if appropriate)
- [x] updates `CHANGELOG.md` (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

Note that your PR will not be reviewed unless all three boxes are checked.

## What does this PR change?
- Reduces the docker image size from ~900MB to ~490MB by leveraging the `python:*-slim` images (addressing #1638 ).
- Introduces a prefect docker image for each supported python version (3.5, 3.6, and 3.7) which are built on each tagged master commit.

## Why is this PR important?
This cuts down the time it takes to pull down the prefect image for the first time, thus getting users to their first flow faster.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -266,41 +267,74 @@ jobs:
...
           name: Test image
           command: |
             docker run -dit prefecthq/prefect /bin/bash -c 'curl -fL0 https://raw.githubusercontent.com/PrefectHQ/prefect/master/examples/retries_with_mapping.py | python'
       - run:
-          name: Authenticate with Docker Hub and push
+          name: Push versioned tags
           command: |
             docker login --username  $DOCKER_HUB_USER --password $DOCKER_HUB_PW
-            docker push prefecthq/prefect
+            docker push prefecthq/prefect:${CIRCLE_TAG}-${PYTHON_TAG}
```
Comment (cicdw) [inline .circleci/config.yml]: I think we might want to finesse the logic here a little, because I suspect this will end up pushing an image with a messy git tag

File: Dockerfile [hunk 2]
```diff
@@ -1,8 +1,20 @@
...
+ARG BUILD_DATE
+
+LABEL maintainer="help@prefect.io"
+LABEL io.prefect.python-version=${PYTHON_VERSION}
+LABEL org.label-schema.schema-version = "1.0"
+LABEL org.label-schema.name="prefect"
+LABEL org.label-schema.url="https://www.prefect.io/"
+LABEL org.label-schema.version=${PREFECT_VERSION}
+LABEL org.label-schema.vcs-ref=${GIT_SHA}
+LABEL org.label-schema.build-date=${BUILD_DATE}
```
Comment (cicdw) [inline Dockerfile]: Nice!

File: src/prefect/environments/storage/docker.py [hunk 3]
```diff
@@ -79,19 +72,50 @@ def __init__(
...
+
         self.env_vars = env_vars or {}
+        self.env_vars["PREFECT__USER_CONFIG_PATH"] = "/root/.prefect/config.toml"
+
         self.files = files or {}
         self.flows = dict()  # type: Dict[str, str]
         self._flows = dict()  # type: Dict[str, "prefect.core.flow.Flow"]
         self.base_url = base_url or default_url
         self.local_image = local_image
+        self.extra_commands = []
```
Comment (cicdw) [inline src/prefect/environments/storage/docker.py]: This feels like a really convenient abstraction - I'd love let's expose this as a new keyword argument for `Docker` storage.

Comment (wagoodman) [reply src/prefect/environments/storage/docker.py]: Maybe, but the ultimate abstraction to this is the `base_image` option where you can run anything that you'd like. In this way we reduce the risk of breaking anyone downstream when we make any overhauls to the base image or reorder where the commands are run.

File: .circleci/config.yml (hunk 1 above)
Comment (wagoodman) [reply .circleci/config.yml]: I'm open to a better way to do it. I added [circle-ci git tag filters here](https://github.com/PrefectHQ/prefect/pull/1648/files/3aad2e7a686429f40856e074712a1dd4d3fc9d0e#diff-1d37e48f9ceff6d8030570cd36286a61R354-R355) to help reduce the risk of an odd tag coming through.

Comment (cicdw) [reply .circleci/config.yml]: Hmmm yea but won't this job still run on master?

File: src/prefect/environments/storage/docker.py (hunk 3 above)
Comment (cicdw) [reply src/prefect/environments/storage/docker.py]: That's true, but many of users aren't super familiar with Docker - this is also the reason we exposed `files` and `python_dependencies` as keywords, to help optionally customize the image without building a base image yourself.  But the risk of breakage is a good point, happy to leave this as-is.

File: src/prefect/environments/storage/docker.py [hunk 4]
```diff
@@ -79,19 +72,50 @@ def __init__(
...
             self.prefect_version = "master" if len(version) > 1 else version[0]
         else:
             self.prefect_version = prefect_version
 
+        if base_image is None:
+            python_version = "{}.{}".format(
+                sys.version_info.major, sys.version_info.minor
+            )
+            if re.match("^[0-9]+\.[0-9]+\.[0-9]+$", self.prefect_version) != None:
+                # note: this does not necessarily mean that we have built/pushed all previous prefect versions to dockerhub
```
Comment (cicdw) [inline src/prefect/environments/storage/docker.py]: luckily a user will only hit this block of code if they're running 0.6.8+ so we don't have to retroactively build old images with this pattern

File: src/prefect/environments/storage/docker.py [hunk 5]
```diff
@@ -79,19 +72,50 @@ def __init__(
         self.image_name = image_name
         self.image_tag = image_tag
         self.python_dependencies = python_dependencies or []
+        self.python_dependencies.append("wheel")
+
         self.env_vars = env_vars or {}
+        self.env_vars["PREFECT__USER_CONFIG_PATH"] = "/root/.prefect/config.toml"
```
Comment (cicdw) [inline src/prefect/environments/storage/docker.py]: Just out of curiosity why was this necessary?  I believe this should be the default.

Comment (cicdw) [reply src/prefect/environments/storage/docker.py]: Nvm I now see that you just moved this from where it previously was 👍

File: tests/environments/storage/test_docker_storage.py [hunk 6]
```diff
@@ -47,11 +47,13 @@ def test_empty_docker_storage(monkeypatch, platform, url):
...
     assert not storage.registry_url
-    assert storage.base_image.startswith("python:")
+    assert storage.base_image.startswith("prefecthq/prefect:python")
     assert not storage.image_name
     assert not storage.image_tag
-    assert not storage.python_dependencies
-    assert not storage.env_vars
+    assert storage.python_dependencies == ["wheel"]
+    assert storage.env_vars == {
+        "PREFECT__USER_CONFIG_PATH": "/root/.prefect/config.toml"
```
Comment (cicdw) [inline tests/environments/storage/test_docker_storage.py]: @joshmeek do you remember why we had to include this originally? I suspect it's no longer necessary but I don't want to take any chances

Comment (joshmeek) [reply tests/environments/storage/test_docker_storage.py]: I don't think this is used anymore

Comment (TylerWanner) [reply tests/environments/storage/test_docker_storage.py]: don't remove it as part of this PR, but do remove it as part of another PR after this gets merged pls

File: .circleci/config.yml (hunk 1 above)
Comment (wagoodman) [reply .circleci/config.yml]: good call, I thought these were `and`ed instead of `or`ed. I'll update...

File: src/prefect/environments/storage/docker.py (hunk 3 above)
Comment (wagoodman) [reply src/prefect/environments/storage/docker.py]: I think I'll leave it as is for now, it doesn't prevent us from exposing this in the future.

File: src/prefect/environments/storage/docker.py (hunk 4 above)
Comment (wagoodman) [reply src/prefect/environments/storage/docker.py]: true, but this won't stop them from specifying an older version [here](https://github.com/PrefectHQ/prefect/blob/7576d7ba0b9c813d9af2004a67c74ef7a458a60d/src/prefect/environments/storage/docker.py#L62) (such as 0.6.6) when using 0.6.8+ locally. I don't see why someone would want to do this, but they could I suppose. 

Though, I take your larger point... I'm willing to let the docker pull fail like we would do for any other user-specified `base_image`. I'll take out the code comment.

File: .circleci/config.yml [hunk 7]
```diff
@@ -266,41 +267,74 @@ jobs:
           name: Upload Coverage
           command: bash <(curl -s https://codecov.io/bash) -cF python -s "/tmp/workspace/coverage/"
 
-  build_image:
+  build-docker-image:
```
[the remark this answers is not in the document]
Comment (wagoodman) [reply .circleci/config.yml]: righto, just swapped over all job names and parameters to underscore nomenclature

File: .circleci/config.yml (hunk 1 above)
Comment (wagoodman) [reply .circleci/config.yml]: Interestingly enough there isn't a good way to do this, @TylerWanner and I came up with this combo:
- https://github.com/PrefectHQ/prefect/pull/1648/files#diff-1d37e48f9ceff6d8030570cd36286a61R361-R364
- https://github.com/PrefectHQ/prefect/pull/1648/files#diff-1d37e48f9ceff6d8030570cd36286a61R286-R292

Up to a better approach, but this is all I've got until circleci supports and'ed filters.

File: tests/environments/storage/test_docker_storage.py [hunk 8]
```diff
@@ -362,9 +387,20 @@ def test_create_dockerfile_from_everything():
             with open(os.path.join(tempdir, "Dockerfile"), "r") as dockerfile:
                 output = dockerfile.read()
 
+            print(output)
+
```
Comment (joshmeek) [inline tests/environments/storage/test_docker_storage.py]:
```suggestion

```

File: .circleci/config.yml [hunk 9]
```diff
@@ -266,43 +267,85 @@ jobs:
...
+      - setup_remote_docker:
+          docker_layer_caching: true
+      - run:
+          name: Build image
+          command: >-
+            docker build 
+            --build-arg GIT_SHA=$CIRCLE_SHA1 
+            --build-arg BUILD_DATE=$(date -u +'%Y-%m-%dT%H:%M:%SZ')
+            --build-arg PREFECT_VERSION=$CIRCLE_TAG 
+            -t prefecthq/prefect:${CIRCLE_TAG}-${PYTHON_TAG}
```
Comment (cicdw) [inline .circleci/config.yml]: whenever `CIRCLE_TAG` is not populated (e.g., on a normal merge-to-master), will this be a valid image tag?

File: .circleci/config.yml (hunk 1 above)
Comment (cicdw) [reply .circleci/config.yml]: I think this is the only place I'm still hesitant; here's what I imagine happening:
- a PR is merged to master
- `build_docker_image` is run
- `CIRCLE_TAG` is empty, because this isn't a tagged commit
- either this step or the build step errors out in some unique way based on this

Everything else looks great though! And to be honest, we could always clean things up if that happens but I want to make sure I understand this piece

File: .circleci/config.yml (hunk 9 above)
Comment (wagoodman) [reply .circleci/config.yml]: no it will not, I believe it would be an empty or missing value. That is the reason for [the filter](https://github.com/PrefectHQ/prefect/pull/1648/files#diff-1d37e48f9ceff6d8030570cd36286a61R364), such that we are only calling this job upon a tagged commit.

However, I could add a `set -u` before this command such that bash will throw an error upon attempting to access a non-existent or empty variable?

File: .circleci/config.yml (hunk 1 above)
Comment (wagoodman) [reply .circleci/config.yml]: I don't believe that it is possible to reach this step because of the workflow filter, however, I'm all for adding the precautionary `set -u` mentioned in the other comment below.
